dlt/common/storages/transactions.py

```diff
@@ -0,0 +1,223 @@
+
+import fsspec
+
+
+def lock_id(k: int = 4) -> str:
+    """Generate a time based random id.
+
+    Args:
+        k: The length of the suffix after the timestamp.
+
+    Returns:
+        A time sortable uuid.
+    """
+    suffix = "".join(random.choices(string.ascii_lowercase, k=k))
+    return f"{time.time_ns()}{suffix}"
+
+
+
+class Heartbeat(Timer):
+    """A thread designed to periodically execute a fn."""
+    daemon = True
+
+    def run(self) -> None:
+        while not self.finished.wait(self.interval):
+            self.function(*self.args, **self.kwargs)
+        self.finished.set()
+
+
+class TransactionalFile:
+    """A transaction handler which wraps a file path."""
+
+    # Map of protocol to mtime resolver
+    # we only need to support a small finite set of protocols
+    mtime_dispatch = {
+        "s3": lambda f: datetime.strptime(f["LastModified"], "%Y-%m-%d %H:%M:%S%z"),
+        "gcs": lambda f: datetime.strptime(f["updated"], "%Y-%m-%dT%H:%M:%S.%fZ"),
+        "adl": lambda f: datetime.strptime(f["LastModified"], "%Y-%m-%d %H:%M:%S%z"),
+        "file": lambda f: datetime.fromtimestamp(f["mtime"]),
+    }
+    # Support aliases
+    mtime_dispatch["gs"] = mtime_dispatch["gcs"]
+    mtime_dispatch["s3a"] = mtime_dispatch["s3"]
+    mtime_dispatch["azure"] = mtime_dispatch["adl"]
+
+    POLLING_INTERVAL = 0.5
+    LOCK_TTL_SECONDS = 30.0
+
+    def __init__(self, path: str, fs: fsspec.AbstractFileSystem) -> None:
+        """Creates a new FileTransactionHandler.
+
+        Args:
+            path: The path to lock.
+            fs: The fsspec file system.
+        """
+        self.path = path
+        self.lock_prefix = f"{self.path}.lock"
+        self.lock_path = f"{self.lock_prefix}.{lock_id()}"
+        proto = fs.protocol[0] if isinstance(fs.protocol, (list, tuple)) else fs.protocol
+        self.get_mtime = self.mtime_dispatch.get(proto, self.mtime_dispatch["file"])
+        self._fs = fs
+        self._original_contents: t.Optional[bytes] = None
+        self._is_locked = False
+        self._heartbeat: t.Optional[Heartbeat] = None
+
+    def _start_heartbeat(self) -> Heartbeat:
+        """Create a thread that will periodically update the mtime."""
+        self._stop_hearbeat()
+        self._heartbeat = Heartbeat(
+            TransactionalFile.LOCK_TTL_SECONDS / 2,
+            self._fs.touch,
+            args=(self.lock_path,),
+        )
+        self._heartbeat.start()
+        return self._heartbeat
+
+    def _stop_hearbeat(self) -> None:
+        """Stop the heartbeat thread if it exists."""
+        if self._heartbeat is not None:
+            self._heartbeat.cancel()
+            self._heartbeat = None
+
+    def _sync_locks(self) -> t.List[str]:
+        """Gets a map of lock paths to their last modified times and removes stale locks."""
+        output = []
+        now = datetime.now()
+
+        for lock in self._fs.glob(self.lock_prefix + ".*", refresh=True, detail=True).values():
```

Comment:
@z3z1ma I've just went through fsspec glob implementation because it often returned no results when I was globbing in one process and deleting files from the same "folder" in another. the implementation is quite complicated and I still do not know why the problem happens. it surely tries to do recursive walk (which is normal for glob) which is not necessary here.
I ended up using `ls` and filtering by prefix myself. I problems gone. I think you should do the same here